File: apps/ios/LanjingQuiz/Models/BankQuestion.swift
import Foundation

/// One practice question. Crawled from upstream via PracticeMapping.bankQuestion
/// and persisted as a JSONL line per category (same on-disk format as the
/// collector's apps/bank/data), so it is Codable both ways: the answer encodes
/// as a single letter string ("A"), a letter array (["A","C"]) or null —
/// matching the collector's record format. Formula images stay remote — only
/// protocol-relative srcs are normalized at construction/decoding time.
struct BankQuestion: Identifiable, Equatable, Codable, Sendable {

    /// answer: "A" | ["A","C"] | null. null → the record has no known answer.
    struct Answer: Codable, Equatable, Sendable {
        let letters: [String]

        init(letters: [String]) {
            self.letters = letters
        }

        init(from decoder: Decoder) throws {
            let container = try decoder.singleValueContainer()
            if let single = try? container.decode(String.self) {
                letters = [single]
            } else if let array = try? container.decode([String].self) {
                letters = array
            } else {
                // JSON null decodes as nil at the answer field (try? at the
                // call site); any other shape is a malformed record.
                throw DecodingError.typeMismatch(
                    Answer.self,
                    DecodingError.Context(
                        codingPath: decoder.codingPath,
                        debugDescription: "answer must be a string, an array of strings, or null"
                    )
                )
            }
        }

        func encode(to encoder: Encoder) throws {
            var container = encoder.singleValueContainer()
            if letters.count == 1 {
                try container.encode(letters[0])
            } else {
                try container.encode(letters)
            }
        }
    }

    let id: String // _id
    let category: String
    let section: String // coarser group, e.g. 逻辑填空 (kept for metadata display)
    let subCategory: String // 题型细分, e.g. 加强支持 — the practice grouping key
    let question: String // HTML, img srcs normalized at construction time
    /// Shared material for comb (资料分析) questions; nil/absent in records
    /// crawled before stems were stored.
    let stem: String?
    let options: [String] // all 4 slots preserved; empty strings = 填空 slots
    let answer: Answer?
    let analysis: String?
    let sourceExamName: String?
    let round: Int?
    let collectedAt: String? // raw ISO8601 string, no date parsing needed

    /// "ABCD" prefix matching the preserved option slots.
    var letters: [String] {
        Array("ABCD".prefix(options.count)).map(String.init)
    }

    /// Correct-flag per letter slot (all false when the answer is unknown).
    var keys: [Bool] {
        letters.map { answer?.letters.contains($0) ?? false }
    }

    /// More than one correct letter (the option count is always 4).
    var isMulti: Bool { (answer?.letters.count ?? 0) > 1 }
    var correctAnswers: [String] { answer?.letters ?? [] }
    var isGradable: Bool { answer != nil }

    /// Protocol-relative img srcs ("//host/...") → "https://host/...".
    /// Absolute https URLs (fb.fbstatic.cn, test.lanjingweike.com) pass
    /// through unchanged. Current bank data has no http:// srcs; if future
    /// collection introduces them, WKWebView's mixed-content policy would
    /// block them (document base is https) — only "//" is repaired here.
    static func normalizeImgSrcs(_ html: String) -> String {
        html
            .replacingOccurrences(of: "src=\"//", with: "src=\"https://")
            .replacingOccurrences(of: "src='//", with: "src='https://")
    }

    init(
        id: String, category: String, section: String, subCategory: String,
        question: String, stem: String?, options: [String], answer: Answer?,
        analysis: String?, sourceExamName: String?, round: Int?, collectedAt: String?
    ) {
        self.id = id
        self.category = category
        self.section = section
        self.subCategory = subCategory
        self.question = Self.normalizeImgSrcs(question)
        self.stem = stem.map(Self.normalizeImgSrcs)
        self.options = options
        self.answer = answer
        self.analysis = analysis.map(Self.normalizeImgSrcs)
        self.sourceExamName = sourceExamName
        self.round = round
        self.collectedAt = collectedAt
    }

    private enum CodingKeys: String, CodingKey {
        case id = "_id"
        case category
        case section
        case subCategory
        case question
        case stem
        case options
        case answer
        case analysis
        case sourceExamName
        case round
        case collectedAt
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.init(
            id: try container.decode(String.self, forKey: .id),
            category: (try? container.decode(String.self, forKey: .category)) ?? "",
            section: (try? container.decode(String.self, forKey: .section)) ?? "",
            subCategory: (try? container.decode(String.self, forKey: .subCategory)) ?? "",
            question: (try? container.decode(String.self, forKey: .question)) ?? "",
            stem: try? container.decode(String.self, forKey: .stem),
            options: (try? container.decode([String].self, forKey: .options)) ?? [],
            answer: try? container.decode(Answer.self, forKey: .answer),
            analysis: try? container.decode(String.self, forKey: .analysis),
            sourceExamName: try? container.decode(String.self, forKey: .sourceExamName),
            round: try? container.decode(Int.self, forKey: .round),
            collectedAt: try? container.decode(String.self, forKey: .collectedAt)
        )
    }
}

File: apps/android/LanjingQuiz/src/main/java/com/qzh/lanjingquiz/Data/BankQuestion.kt
package com.qzh.lanjingquiz.Data

import com.qzh.lanjingquiz.Support.HtmlHelpers
import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.descriptors.buildClassSerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonDecoder
import kotlinx.serialization.json.JsonEncoder
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray

/**
 * 答案三态(iOS BankQuestion.Answer 移植):"A"(单选)/ ["A","C"](多选)/ null(无答案)。
 * 解码容错:字符串 → Single(恰 1 字符,否则 None);数组 → Multi;null/缺失/其他 → None。
 * 编码:Single → 字符串(绝不一元素数组);Multi → 数组;None → null。
 */
@Serializable(with = AnswerShape.Serializer::class)
sealed interface AnswerShape {
    data class Single(val letter: String) : AnswerShape
    data class Multi(override val letters: List<String>) : AnswerShape
    data object None : AnswerShape

    /** 判分用字母列表;None → 空。 */
    val letters: List<String>
        get() = when (this) {
            is Single -> listOf(letter)
            is Multi -> letters
            None -> emptyList()
        }

    object Serializer : KSerializer<AnswerShape> {
        override val descriptor: SerialDescriptor = buildClassSerialDescriptor("AnswerShape")

        override fun serialize(encoder: Encoder, value: AnswerShape) {
            val json = encoder as? JsonEncoder ?: error("AnswerShape only works with JSON")
            json.encodeJsonElement(when (value) {
                is Single -> JsonPrimitive(value.letter)
                is Multi -> buildJsonArray { value.letters.forEach { add(JsonPrimitive(it)) } }
                None -> JsonNull
            })
        }

        override fun deserialize(decoder: Decoder): AnswerShape {
            val json = decoder as? JsonDecoder ?: error("AnswerShape only works with JSON")
            return when (val el = json.decodeJsonElement()) {
                is JsonPrimitive ->
                    if (el.isString && el.content.length == 1) Single(el.content) else None
                is JsonArray ->
                    Multi(el.mapNotNull { (it as? JsonPrimitive)?.takeIf { p -> p.isString }?.content })
                else -> None
            }
        }
    }
}

/**
 * 一条练习题目(JSONL 一行;iOS BankQuestion 移植)。JSON 键逐字 spec §3.2:
 * _id/category/section/subCategory/question/stem/options/answer/analysis/
 * sourceExamName/round/collectedAt;options 恒 4 槽,空串 = 填空。
 * question/stem/analysis 的协议相对图片 src 在构造/读取路径统一归一化。
 */
@Serializable
data class BankQuestion(
    @SerialName("_id") val id: String,
    val category: String = "",
    val section: String = "",
    @SerialName("subCategory") val subCategory: String = "",
    val question: String = "",
    val stem: String? = null,
    val options: List<String> = emptyList(),
    val answer: AnswerShape? = null,
    val analysis: String? = null,
    @SerialName("sourceExamName") val sourceExamName: String? = null,
    val round: Int? = null,
    @SerialName("collectedAt") val collectedAt: String? = null,
)

/** 读取路径统一归一化(协议相对图片 src → https),与 iOS init(from:) 解码归一行为一致。 */
internal fun BankQuestion.normalized(): BankQuestion = copy(
    question = HtmlHelpers.normalizeImgSrcs(question) ?: question,
    stem = HtmlHelpers.normalizeImgSrcs(stem),
    analysis = HtmlHelpers.normalizeImgSrcs(analysis),
)
